src/proofs/weighted_inner_product.rs

```diff
@@ -0,0 +1,915 @@
+
+            let yn_aR = (0..n)
+                .map(|i| BigInt::mod_mul(&powers_y[n - 1], &a_R[i], &order))
+                .collect::<Vec<BigInt>>();
+            let yninv_aL = (0..n)
+                .map(|i| BigInt::mod_mul(&powers_yinv[n - 1], &a_L[i], &order))
+                .collect::<Vec<BigInt>>();
+
+            let c_L = weighted_inner_product(&a_L, &b_R, y.clone());
+            let c_R = weighted_inner_product(&yn_aR, b_L, y.clone());
+
+            // Note that no element in vectors a_L and b_R can be 0
+            // since 0 is an invalid secret key!
+            //
+            // L = <yninv_aL * G_R> + <b_R * H_L> + (c_L * g) + (d_L * h)
+            let c_L_fe: FE = ECScalar::from(&c_L);
+            let g_cL: GE = g * &c_L_fe;
+            let d_L_fe: FE = ECScalar::new_random();
+            let h_dL = h * &d_L_fe;
+            let g_cL_h_dL = g_cL.add_point(&h_dL.get_element());
+            let yninv_aL_GR = G_R.iter().zip(yninv_aL.clone()).fold(g_cL_h_dL, |acc, x| {
+                if x.1 != BigInt::zero() {
+                    let aLi: FE = ECScalar::from(&x.1);
+                    let aLi_GRi: GE = x.0 * &aLi;
+                    acc.add_point(&aLi_GRi.get_element())
+                } else {
+                    acc
+                }
+            });
+            let L = H_L.iter().zip(b_R.clone()).fold(yninv_aL_GR, |acc, x| {
+                if x.1 != &BigInt::zero() {
+                    let bRi: FE = ECScalar::from(&x.1);
+                    let bRi_HLi: GE = x.0 * &bRi;
+                    acc.add_point(&bRi_HLi.get_element())
+                } else {
+                    acc
+                }
+            });
+
+            // Note that no element in vectors a_R and b_L can be 0
+            // since 0 is an invalid secret key!
+            //
+            // R = <yn_aR * G_R> + <b_R * H_L> + (c_R * g) + (d_R * h)
+            let c_R_fe: FE = ECScalar::from(&c_R);
+            let g_cR: GE = g * &c_R_fe;
+            let d_R_fe: FE = ECScalar::new_random();
+            let h_dR = h * &d_R_fe;
+            let g_cR_h_dR = g_cR.add_point(&h_dR.get_element());
+            let aR_GL = G_L.iter().zip(yn_aR.clone()).fold(g_cR_h_dR, |acc, x| {
+                if x.1 != BigInt::zero() {
+                    let aRi: FE = ECScalar::from(&x.1);
+                    let aRi_GLi: GE = x.0 * &aRi;
+                    acc.add_point(&aRi_GLi.get_element())
+                } else {
+                    acc
+                }
+            });
+            let R = H_R.iter().zip(b_L.clone()).fold(aR_GL, |acc, x| {
+                if x.1 != &BigInt::zero() {
+                    let bLi: FE = ECScalar::from(&x.1);
+                    let bLi_HRi: GE = x.0 * &bLi;
+                    acc.add_point(&bLi_HRi.get_element())
+                } else {
+                    acc
+                }
+            });
+
+            let x = HSha256::create_hash_from_ge(&[&L, &R, &g, &h]);
```

Comment:
why `x` and not `e` ?

Reply:
For pre-final rounds, I am using `x` as the challenge so as to avoid confusion with the final-round challenge `e`. The confusion can arise if I use `e` in both places in the multi-exponentiation verification function.